cmd/util/cmd/epoch-recovery/cmd/reset.go

```diff
@@ -0,0 +1,213 @@
+)
+
+// resetCmd represents a command to reset epoch data in the Epoch smart contract
+var resetCmd = &cobra.Command{
+	Use:   "reset",
+	Short: "Generates `resetEpoch` JSON transaction arguments",
+	Long:  "Generates `resetEpoch` transaction arguments from a root protocol state snapshot and writes it to a JSON file",
+	Run:   resetRun,
+}
+
+func init() {
+	rootCmd.AddCommand(resetCmd)
+	addResetCmdFlags()
+}
+
+func addResetCmdFlags() {
+	resetCmd.Flags().StringVar(&flagBootDir, "boot-dir", "", "path to the directory containing the bootstrap files")
+	_ = resetCmd.MarkFlagRequired("boot-dir")
+
+	resetCmd.Flags().StringVar(&flagPayout, "payout", "", "the payout eg. 10000.0")
+}
+
+// resetRun resets epoch data in the Epoch smart contract with fields generated
+// from the root-protocol-snapshot.json
+func resetRun(cmd *cobra.Command, args []string) {
+
+	// path to the root protocol snapshot json file
+	snapshotPath := filepath.Join(flagBootDir, bootstrap.PathRootProtocolStateSnapshot)
+
+	// check if root-protocol-snapshot.json file exists under the dir provided
+	exists, err := pathExists(snapshotPath)
+	if err != nil {
+		log.Fatal().Err(err).Str("path", snapshotPath).Msgf("could not check if root protocol-snapshot.json exists")
+	}
+	if !exists {
+		log.Error().Str("path", snapshotPath).Msgf("root-protocol-snapshot.json file does not exists in the --boot-dir given")
+		return
+	}
+
+	// path to the JSON encoded cadence arguments for the `resetEpoch` transaction
+	path, err := os.Getwd()
+	if err != nil {
+		log.Fatal().Err(err).Msgf("could not get working directory path")
+	}
+	argsPath := filepath.Join(path, resetArgsFileName)
+
+	// read root protocol-snapshot.json
+	bz, err := io.ReadFile(snapshotPath)
+	if err != nil {
+		log.Fatal().Err(err).Msgf("could not read root snapshot file")
+	}
+	log.Info().Str("snapshot_path", snapshotPath).Msg("read in root-protocol-snapshot.json")
+
+	// unmarshal bytes to inmem protocol snapshot
+	snapshot, err := convert.BytesToInmemSnapshot(bz)
+	if err != nil {
+		log.Fatal().Err(err).Msg("could not convert array of bytes to snapshot")
+	}
+
+	// extract arguments from reset epoch tx from snapshot
+	txArgs := extractResetEpochArgs(snapshot)
+	log.Info().Msg("extracted resetEpoch transaction arguments from snapshot")
+
+	// ancode to JSON
+	enc := encodeArgs(txArgs)
+
+	// write JSON args to file
+	err = io.WriteFile(argsPath, enc)
+	if err != nil {
+		log.Fatal().Err(err).Msg("could not write jsoncdc encoded arguments")
+	}
+	log.Info().Str("path", argsPath).Msg("wrote resetEpoch transaction arguments")
+}
+
+// extractResetEpochArgs extracts the required transaction arguments for the `resetEpoch` transaction
+func extractResetEpochArgs(snapshot *inmem.Snapshot) []cadence.Value {
+
+	// get current epoch
+	epoch := snapshot.Epochs().Current()
+
+	// read epoch counter
+	epochCounter, err := epoch.Counter()
+	if err != nil {
+		log.Fatal().Err(err).Msg("could not get epoch counter")
+	}
```

Comment:
The `epochCounter` value expected by the smart contract is the epoch _being replaced_, which is one less than the epoch beginning after the spork. See [this line](https://github.com/onflow/flow-core-contracts/blob/feature/epochs/contracts/epochs/FlowEpoch.cdc#L436) in `resetEpoch`.

We need to pass in the counter for the epoch being replaced (or "skipped") in the smart contract (ie. `epochCounter-1` and we should also add some comments explaining why this is, as it's a bit tricky.

Reply:
Done.